Add source location and comments to assertions generated from function contracts

feliperodri opened the pull request:
Resolves #5526.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -876,6 +876,9 @@ void code_contractst::add_contract_check(
     convert_to_goto(
       code_assertt(ensures_cond), function_symbol.mode, assertion);
+    assertion.instructions.front().source_location.set_line(ensures.source_location().get_line());
+    assertion.instructions.back().source_location.set_comment("ensures clause");
```
What about the file name?

feliperodri replied:
Yep, I added it.

martin-cs reviewed (approved):
Is this the only place where something like this happens?  It feels like there are likely to be other cases of this in the code contract implementation.

feliperodri committed:
Adds source location to function contracts

Function contracts inject assertions into the goto program to map
pre- and postconditions. These injected assertions must include
original source-code location and meaningful comments for usability.

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

SaswatPadhi commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -436,11 +436,17 @@ bool code_contractst::apply_function_contract(
     code_contractst::add_quantified_variable(requires, replace, mode);
     replace(requires);
 
+    //goto_programt assertion;
```
Should this be removed?

feliperodri replied:
Yes.

SaswatPadhi reviewed (commented):
As part of this PR, could we also add source location and comments to assigns clause? (#5526)

SaswatPadhi commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -563,7 +568,8 @@ void code_contractst::instrument_assign_statement(
     instruction_iterator->source_location));
-
+  alias_assertion.instructions.back().source_location.set_comment(
+    "Check assigns clause");
```
I think the `Check ensures clause` and `Check requires clause` messages are fine, because each function currently can have at most one ENSURES clause and at most one REQUIRES clause.

But multiple `Check assigns clause` assertions seem less useful because it's not clear which variable each assertion is looking at.

There might be a better way to do this, but this could be one way:

```cpp
alias_assertion.instructions.back().source_location.set_comment(
    "Check that " + lhs.pretty(0) + " is assignable");
```

@tautschnig Do you have any suggestions on what we should use here to `string`-ify `lhs` here? Is `pretty` the right thing to use?

SaswatPadhi replied:
As we discussed offline, `pretty` seems to generate a verbose string (with lots of debug info).

Based on a quick search through the code base, I think `from_expr` (from `#include <langapi/language_util.h>`) could be the used here, e.g. https://github.com/diffblue/cbmc/blob/1ab5de1ac0893c7ca94ffe5463e2c7c4ee781266/src/goto-programs/builtin_functions.cpp#L752-L753

feliperodri replied:
@SaswatPadhi I also tried `from_expr`, but at the end it also uses pretty and generates the same verbose strings.
https://github.com/diffblue/cbmc/blob/1ab5de1ac0893c7ca94ffe5463e2c7c4ee781266/src/langapi/language.cpp#L32-L39

I'm now converting `lhs` to a symbol and extracting the identifier, which results to the following message
```console
main.c function bar
[bar.1] line 4 Check that bar::y is assignable: SUCCESS
```

SaswatPadhi replied:
The above `from_expr` definition looks suspicious, I don't think that's the one being called from the `set_comment` example I showed.

The `from_expr` call should return a `std::string` but the definition you posted above returns a `bool`. Could you please double check?

It's being used within `set_comment`, so if it prints a verbose output then that should break the CBMC output ...

SaswatPadhi replied:
I believe this is the right definition:

https://github.com/diffblue/cbmc/blob/48893287099cb5780302fe9dc415eb6888354fd6/src/langapi/language_util.cpp#L20-L31

Also try calling:
https://github.com/diffblue/cbmc/blob/48893287099cb5780302fe9dc415eb6888354fd6/src/langapi/language_util.cpp#L59-L63

The problem with your "symbol extraction" approach would be that not all assignment targets would be symbols. We also have derefs and struct members, for example.

SaswatPadhi replied:
Oh may be you're right, `p->from_expr` probably resolves to the definition you mentioned ... 

If that's the case, then my last suggestion would be to try `expr2c` function from `#include <ansi-c/expr2c.h>`. 

If that doesn't work either, then let's wait to hear from @tautschnig. But I wonder why `from_expr` is used within `set_comments` in the codebase, if it prints a debug output, not a small inline expression ...

SaswatPadhi replied:
Okay, this should be called, not the abstract definition that you posted above.

https://github.com/diffblue/cbmc/blob/1ab5de1ac0893c7ca94ffe5463e2c7c4ee781266/src/ansi-c/ansi_c_language.cpp#L146-L153

This does call `expr2c` and not `pretty`.

Basically in this case the `get_language_from_identifier` should resolve to C language (i.e. `ansi_c_languaget`), not the abstract `languaget`. If that doesn't happen, let's look into the root cause.

feliperodri replied:
Thanks for the tip! It worked perfectly.

SaswatPadhi reviewed (commented):
One comment below.

And please check the assertions introduced at lines 634 and 669. Those don't have any comments, but look related to assigns clause?

feliperodri committed:
Updates contracts regression-test descriptions

Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

SaswatPadhi commented on regression/contracts/assigns_enforce_scoping_01/test.desc:
```diff
@@ -3,8 +3,9 @@ main.c
-assertion: FAILURE$
+^\[f1.\d+\] line \d+ Check that b is assignable: SUCCESS$
+^\[f1.\d+\] line \d+ Check that \*b is assignable: SUCCESS$
+^\[f1.\d+\] line \d+ Check that \*b is assignable: FAILURE$
```
Were we generating 3 assertions before this change as well?

The function never assigns to `b`, only to `*b`, so why are we checking:
```
^\[f1.\d+\] line \d+ Check that b is assignable: SUCCESS$
```

Is it because of this line:

```c
int *b = a;
```

If so, is it necessary to check assigns clause at local variable creation? I mean they would never violate the assigns clause right? And probably shouldn't even be checked, just for efficiency.

If we agree and I'm not missing something here, then let's create a ticket for this and fix at some point?

feliperodri replied:
We currently check ALL assignments in the function body, even if they are over local variables. I agree with you that this is unnecessary and we could prune these cases for efficiency. See https://github.com/diffblue/cbmc/issues/6154

SaswatPadhi replied:
Great, thanks for creating the issue. Resolving this comment.

SaswatPadhi commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -563,7 +568,11 @@ void code_contractst::instrument_assign_statement(
-
+  std::string source_expr_string;
+  (get_language_from_identifier(ns, lhs.id()))
+    ->from_expr(lhs, source_expr_string, ns);
```
These 3 lines look like we are inlining the definition of https://github.com/diffblue/cbmc/blob/48893287099cb5780302fe9dc415eb6888354fd6/src/langapi/language_util.cpp#L20-L31

Could you please explain what is the issue with doing the following (after `#include`ing the appropriate headers):
```c
    "Check that " + from_expr(ns, lhs.id(), lhs) + " is assignable"
```

feliperodri replied:
Perfect! Implemented as suggested.

SaswatPadhi commented on regression/contracts/quantifiers-forall-ensures-enforce/test.desc:
```diff
@@ -3,7 +3,10 @@ main.c
+^\[postcondition.\d+\] file main.c line \d+ Check ensures clause: SUCCESS$
+^\[f1.\d+\] line \d+ Check that i is assignable: SUCCESS$
+^\[f1.\d+\] line \d+ Check that i is assignable: SUCCESS$
+^\[f1.\d+\] line \d+ Check that arr\[\(signed long int\)i\] is assignable: SUCCESS$
```
Ah on Windows this is `signed long long int`:

```
[f1.2] line 18 Check that arr[(signed long long int)i] is assignable: SUCCESS
```

I would just make this:

```
^\[f1.\d+\] line \d+ Check that arr\[\(.*\)i\] is assignable: SUCCESS$
```

feliperodri replied:
Fixed it!

SaswatPadhi reviewed (approved):
Nice! These comments look great :)
